arviz/utils.py

```diff
@@ -50,14 +54,34 @@ def _var_names(var_names, data):
                 )
             )
 
+        excluded_vars = [
+            var[1:] for var in var_names if var.startswith("~") and var not in all_vars
+        ]
         if excluded_vars:
-            var_names = [i for i in all_vars if i not in excluded_vars]
-
-        existent_vars = np.isin(var_names, all_vars)
-        if not np.all(existent_vars):
+            if (filter == "like") or (filter == "regex"):
+                for pattern in excluded_vars:
+                    if pattern not in all_vars:
+                        if filter == "like":
+                            real_vars = [real_var for real_var in all_vars if pattern in real_var]
+                        else:
+                            # i.e filter == "regex"
+                            real_vars = [
+                                real_var for real_var in all_vars if re.search(pattern, real_var)
+                            ]
+                        excluded_vars.extend(real_vars)
+                        excluded_vars.remove(pattern)
```

Comment:
This will skip iterations. If needed, copy first the loop variable

Reply:
I'm not sure I understand -- can you expand a little please?

Reply:
Actually this can blow-up too (infinite iteration). I will tell you when on computer.

(Never edit list while iterating over it)